docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1649,53 +1649,85 @@ fits the encoded integer value. This representation achieves three goals:
-specified bitwidth for a P4 object value, then the server computes the number of
-binary string bytes required to represent the maximum integer of `p4_bitwidth`
-bits as:
-~ Begin CPP
-binary_string_bytes = (p4_bitwidth + 7) / 8;
-~ End CPP
-The server's decision to accept or reject the binary string depends on the
-string's content and its actual length in the P4Runtime request, as follows:
-
-* If the string's actual byte length is shorter than `binary_string_bytes`
-  but greater than zero, the server always accepts the string and treats the
-  missing bytes as zeroes in the most-significant positions.
-
-* If the string's actual byte length is longer than `binary_string_bytes`, the
-  server rejects any string where the extra high-order bytes contain non-zero
-  values. Otherwise, the server truncates the extra bytes from the front
-  of the string and processes the remainder of the string as if it contains
-  exactly `binary_string_bytes`, as described below.
-
-* If the string's original or truncated byte length is equal to
-  `binary_string_bytes`, then the server must also consider the P4 integer
-  type's byte alignment. Byte-aligned strings that satisfy the expression:
-~ Begin CPP
-  (binary_string_bytes * 8 == p4_bitwidth)
-~ End CPP
-  are always valid. For non-aligned strings, the server requires
-  the client to fill the high-order pad bits with zeroes, and the server
-  rejects any binary strings with non-zero pad bits.
-
-* If the string's byte length is zero, the server always rejects the string.
+Any additional bits in the bytes sent for an unsigned integer value
+(type `bit<W>`) must be 0.  If additional bytes are transmitted above
+the `minimum_string_size` minimum required, they must be filled with
+0.
+
+Any additional bits in the bytes sent for a signed integer value (type
+`int<W>`) must be copies of the sign bit, i.e. 0 for non-negative
+values, or 1 for negative values. If additional bytes are transmitted
+above the `minimum_string_size` minimum required, they must be filled
+with copies of the sign bit, i.e. 0 for non-negative values, or 0xff
+for negative values. In 2's complement representation, this is called
+"sign extension", and leaves the numeric value represented unchanged.
+
+Upon receiving a binary string, the P4Runtime receiver (whether the
+server or the client) does not impose any restrictions on the length
+of the string itself. Instead, the receiver verifies that the value
+encoded by the string fits within the expected type (signed or
+unsigned) and P4Info-specified bitwidth for the P4 object value.
+
+For received values expected to fit within a `bit<W>` type, the value
+is in range if, after removing all most significant 0 bits, the
+remaining value's width is `W` bits or less.
+
+For received values expected to fit within an `int<W>` type, the value
+is in range if, after "undoing sign extension", the remaining value's
+width is `W` bits or less. To undo sign extension, start by
```

Comment:
I may be wrong, but isn't it `W-1` bits here?
For example, if I expect `int<2>` (only 1 and -1 can be represented) and the client sends the 1-byte value `00000011b`, then after "undoing sign extension" I get a value (3) that fits in 2 bits but cannot be represented with `int<2>`.

Reply:
The type `int<W>` can always represent `2^W` different values, e.g. `int<2>` can represent any value in the range `[-2, 1]`.

A P4Runtime client/server, if it started with a value that it thought was type `int<2>` represented as `11b`, would _never_ send it as the byte `00000011b` using the proposed mechanism.  It would sign extend the 2-bit value with 1 bits, and send it as `11111111b`.  Both `11b` and `11111111b` represent the numeric value -1 in 2's complement representation.

The only way with the proposed changes that one of the client/server would send `00000011b` as the encoding for a value of type `int<W>` is if the numeric value was +3, and the type would thus have to be `int<3>` or larger at the sender.   If the receiver had type `int<2>` or `int<1>`, i would "un sign extend" the value down to `011b`, but would there have to stop because the 2 most significant bits differ, and reducing it more would change the sign of the value.  It would then detect that the value requires a width of W=3 bits, and since the receiver's type was `int<2>` or `int<1>` in this example scenario, the receiver would declare the value out of range.

Reply:
I can add example scenarios like this, either directly in this section explaining this mechanism, or in an appendix, if you think having such examples would clarify things for readers.  I know that many are not as intimately familiar with the gory details of 2's complement notation as I forced myself to become :)

Reply:
Yes I did forget that P4 used 2's complement for a second there :)
I was also thinking of a sender with `int<3>` and a receiver with `int<2>`, but didn't "unsign extend" properly. My only suggestion would be to maybe use binary string when talking about the bit pattern sent by the client and value for the numerical value. Apart from that I think it's clear.

Reply:
I have made a few changes from "value" to "bit string" in the version pushed to this branch: https://github.com/p4lang/p4runtime/tree/issue111-change-signed-int-representation